### ISSUE TITLE
Structure learning and optimization algs

### ISSUE THREAD

**Issue by anicolson** [SPFlow/SPFlow#51]
Hi, I am using the default parameters for learn_paremetric, with univariate Gaussian leaves. Would someone be able to tell me the structure learning and optimization algorithms used? And if possible, the papers that they are from? Thank you in advance if you are able to help.

-------------------------------------------------------------------------------------------------------------
ds_context = Context(parametric_types=[Gaussian]*M).add_domains(train_batch).

def learn_parametric(
    data,
    ds_context,
    cols="rdc",
    rows="kmeans",
    min_instances_slice=200,
    min_features_slice=1,
    multivariate_leaf=False,
    threshold=0.3,
    ohe=False,
    leaves=None,
    memory=None,
    rand_gen=None,
    cpus=-1,
)
-------------------------------------------------------------------------------------------------------------

**Comment by pranavsubramani** [comment]
The paper that the implementation is from is the following:

1. https://homes.cs.washington.edu/~pedrod/papers/mlc13.pdf

I am a little unsure of what you mean by "optimization in this case" -- parameter learning and structure learning are often different problems, the optimization algorithms are for your "structure-learned-spn", which are:

1. EM (Expectation Maximization)
2. Gradient Descent (Stochastic) -- or pretty much anything that Tensorflow offers.

**Comment by alejandromolinaml** [comment]
Dear Aaron, 

the two parameters control the structure learning algorithm. This algorithm splits rows (using sum nodes) and columns (using product nodes) of your data.
Imagine that you are splitting over and over, eventually, you run out of instances to make any meaningful predictions. 
The parameter:
min_instances_slice=200,
says, split rows unless you have 200 or less instances. In that case, just learn uniform Gaussians (in your case).

To split the columns, we run a series of statistical tests and the threshold is controlled by this parameter.:
threshold=0.3,
It basically controls when we believe groups of variables are independent of each other, this threshold can take values between 0 and 1.

Regarding your problem of finding optimal parameters, this is usually done via splitting your training data into smaller sets of training T and a validation set V. You train on the smaller training T and evaluate the performance on V. 
You do that for a range of values of parameters:
min_instances_slice=[50, 100, 200, ...], threshold=[0.001, 0.01, 0.1, 0.2, 0.3] etc. 
you can decide which range of values might work well for you. 

What Pranav mentioned is that even if you have the structure. You can still use optimization algorithms to find even better parameters for the structure. 
We have implemented the two options he mentioned in SPFLow.

hope this helps, 

Alejandro

**Comment by anicolson** [comment]
Thank you for your help and your time!

________________________________
